### ISSUE TITLE
EventEmitter memory leak on redirect

### ISSUE THREAD

**Issue by blakeembrey** [request/request#1390]
I was debugging https://github.com/blakeembrey/popsicle/issues/10 and found out that there's actually an event listening bug that occurs on every redirect. It keeps attaching the same `pipe` and `end` event listeners every redirect.

I'll try to get a PR with tests ASAP, but didn't want this to get lost.

The referenced issue is somewhat related to https://github.com/request/request/pull/1132 since it hits an infinite redirect on `302` which made this fun to debug. FWIW, `curl` did not get stuck in an infinite redirect (maybe another PR) and correctly closed. 

```
> GET / HTTP/1.1
> User-Agent: curl/7.37.1
> Host: example.com
> Accept: */*
> 
< HTTP/1.1 302 Found
< Date: Fri, 30 Jan 2015 15:12:52 GMT
* Server Apache/2.2.15 (Red Hat) is not blacklisted
< Server: Apache/2.2.15 (Red Hat)
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
< Pragma: no-cache
< Location: http://example.com
< Content-Length: 107
< Connection: close
< Content-Type: text/html; charset=utf-8
< 
* Closing connection 0
<html><head><meta http-equiv="refresh" content="0;url=http://example.com"/></head></html>
```

**Comment by nylen** [comment]
:+1:

Have a look at #1329, that was a decent start but it appears to have been abandoned.

Infinite redirects are always fun :( if you can post the URL/status transitions the request goes through in a separate issue then we can take a look.

**Comment by blakeembrey** [comment]
@nylen It literally just gets stuck in a `302` loop. Curl closes this. Anyway, I'll look at making another PR if that one has stagnated since this should have been caught by tests too (swear it didn't break before).

**Comment by nylen** [comment]
Do we let it delete the cookie?

I had another thought on #1132 which I'm posting there.

**Comment by blakeembrey** [comment]
@nylen I'll check that in the morning, but my hunch is that might be the issue too. I'd guess that the header isn't updated each redirect since the request never ends.

**Comment by jpettitt** [comment]
Did this ever get resolved?  I just hit the issue.  

Does the memory get reclaimed when the `maxRedirects` triggers and the request fails?  I noticed it becasue my max event listeners is set to 10 and maxRedirects was set to 20 which cause a fatal, and uncatchable, error that killed the process.  I've reduced my maxRedirects as a work around but I'd like to be sure I've not just created a memory leak.
